Modules/_abc.c

```diff
@@ -0,0 +1,960 @@
+    if (gs->iterating) {
+        return 0;
+    }
+
+    assert(PyList_CheckExact(gs->pending));
+    Py_ssize_t pos = PyList_GET_SIZE(gs->pending);
+    while (pos > 0) {
+        pos--;
+        PyObject *ref = PyList_GET_ITEM(gs->pending, pos);
+        Py_INCREF(Py_None);
+        PyList_SET_ITEM(gs->pending, pos, Py_None);
+
+        if (ref == Py_None) {
+            Py_DECREF(ref);
+            continue;
+        }
+        if (PySet_Discard(gs->data, ref) < 0) {
+            Py_DECREF(ref);
+            return -1;
+        }
+    }
+    return PyList_SetSlice(gs->pending,
+        0, PyList_GET_SIZE(gs->pending), NULL);
+}
+
+static int
+_add_to_registry(PyObject *self, PyObject *cls)
+{
+    _abc_data *impl;
+    int res;
+    impl = _get_impl(self);
+    if (!impl) {
+        return -1;
+    }
+    res = _add_to_weak_set((PyObject *)(impl->_abc_registry), cls, 1);
+    Py_DECREF(impl);
+    return res;
+}
+
+static int
+_add_to_cache(_abc_data *impl, PyObject *cls)
+{
+    return _add_to_weak_set(impl->_abc_cache, cls, 0);
+}
+
+static int
+_add_to_negative_cache(PyObject *self, PyObject *cls)
+{
+    _abc_data *impl;
+    int res;
+    impl = _get_impl(self);
+    if (!impl) {
+        return -1;
+    }
+    res = _add_to_weak_set(impl->_abc_negative_cache, cls, 0);
+    Py_DECREF(impl);
+    return res;
+}
+
+PyDoc_STRVAR(_reset_registry_doc,
+"Internal ABC helper to reset registry of a given class.\n\
+\n\
+Should be only used by refleak.py");
+
+static PyObject *
+_reset_registry(PyObject *m, PyObject *args)
+{
+    PyObject *self, *registry;
+    if (!PyArg_UnpackTuple(args, "_reset_registry", 1, 1, &self)) {
+        return NULL;
+    }
+    registry = _get_registry(self);
+    if (!registry) {
+        return NULL;
+    }
+    if (PySet_Clear(registry) < 0) {
+        return NULL;
+    }
+    Py_RETURN_NONE;
+}
+
+static int
+_reset_negative_cache(_abc_data *impl)
+{
+    return PySet_Clear(impl->_abc_negative_cache);
+}
+
+PyDoc_STRVAR(_reset_caches_doc,
+"Internal ABC helper to reset both caches of a given class.\n\
+\n\
+Should be only used by refleak.py");
+
+static PyObject *
+_reset_caches(PyObject *m, PyObject *args)
+{
+    PyObject *self;
+    _abc_data *impl;
+    if (!PyArg_UnpackTuple(args, "_reset_caches", 1, 1, &self)) {
+        return NULL;
+    }
+    impl = _get_impl(self);
+    if (!impl) {
+        return NULL;
+    }
+    if (PySet_Clear(impl->_abc_cache) < 0) {
+        Py_DECREF(impl);
+        return NULL;
+    }
+    /* also the second cache */
+    if (_reset_negative_cache(impl) < 0) {
+        Py_DECREF(impl);
+        return NULL;
+    }
+    Py_DECREF(impl);
+    Py_RETURN_NONE;
+}
+
+static PyObject *
+_get_negative_cache_version(_abc_data *impl)
+{
+    return impl->_abc_negative_cache_version;
+}
+
+static void
+_set_negative_cache_version(_abc_data *impl)
+{
+    Py_INCREF(abc_invalidation_counter);
+    Py_SETREF(impl->_abc_negative_cache_version, abc_invalidation_counter);
```

Comment:
This is double incref of `abc_invalidation_counter`. Is that intended?

Reply:
Py_SETREF doesn't incref.  It does only decref.

```
#define Py_SETREF(op, op2)                      \
    do {                                        \
        PyObject *_py_tmp = (PyObject *)(op);   \
        (op) = (op2);                           \
        Py_DECREF(_py_tmp);                     \
    } while (0)
```